docs/src/main/sphinx/develop/supporting-merge.rst

```diff
@@ -0,0 +1,379 @@
+method iterates over the rows in the page, performing updates and deletes
+in the ``MATCHED`` cases, and inserts in the ``NOT MATCHED`` cases.
+
+For some ``RowChangeParadigms``, ``UPDATE`` operations may have been
+translated into the corresponding ``DELETE`` and ``INSERT`` operations
+before ``storeMergedRows(Page)`` was called.
+
+To complete the ``MERGE`` operation, the Trino engine calls
+``ConnectorMetadata.finishMerge(...)``, passing the table
+and that collection of ``Slice`` fragments with information about what
+was changed, and the connector takes appropriate actions, if
+any.
+
+``RowChangeProcessor`` implementation for ``MERGE``
+---------------------------------------------------
+
+In SQL ``MERGE``, each supported ``RowChangeParadigm``
+corresponds to an internal Trino engine class that implements interface
+``RowChangeProcessor``.  ``RowChangeProcessor`` has one interesting method:
+``Page transformPage(Page)``.  The format of the input and output page depend
+on the the SQL operation being implemented.
+
+The connector has no access to the ``RowChangeProcessor`` instance - - it
+is used inside the Trino engine to transform the merge page rows into rows
+to be stored, based on the connector's choice of ``RowChangeParadigm``.
+
+For SQL MERGE, the page supplied to ``transformPage()`` consists of:
+
+* The write redistribution columns if any
+* The ``rowId`` column for the row from the target table if matched, or
+  null if not matched
+* The merge case ``RowBlock``
+* For partitioned or bucketed tables, a hash value column.
+
+The merge case ``RowBlock`` has this layout:
+
+* Blocks for each column in the table, including partition columns, in
+  table column order.
+* A block containing 1 if the target table row was matched by a source
+  row base on the ``MERGE`` ``JOIN`` condition, and 0 if there was not
+  target table row matching the source row using the ``JOIN`` condition.
+  the row starting at 0, or -1 no merge case matched for the row.
+* A block containing the merge case operation, encoded as ``INSERT`` = 1,
+  ``DELETE`` = 2, ``UPDATE`` = 3 and if no merge case matched, -1.
+
+The page returned from ``transformPage`` consists of all table columns,
+in table column order, followed by the rowId column, followed by the
+operation column from the merge case ``RowBlock``.
+
+Interface ``RowChangeProcessor`` now supports SQL MERGE, and will later be
+used to upgrade the SQL UPDATE engine implementation.
+
+Detecting duplicate matching target rows
+----------------------------------------
+
+The SQL ``MERGE`` specification requires that in each ``MERGE`` case,
+a single target table row must match at most one source row, after
+applying the ``MERGE`` case condition expression.  This is done in
+class ``DuplicateRowFinder.checkForDuplicateTargetRows()``.  That method
+examines successive rows produced by the ``SELECT CASE`` expression,
+excluding ``INSERT``ed rows, and if the ``writeRedistributionColumns``
```

Comment:
Changed to "and checks if".

You've made an excellent point about comparing the `writeRedistributionColumns`; I don't believe it's needed.  I'll mull it over and take it up with @electrum.

Reply:
Please let me know the outcome.

Reply:
Actually, the code is correct as it stands: `DuplicateRowFinder` must compare both the `rowId` column and the `writeRedistributionColumns`.  The reason is that the row finder is called before the merge page is separated by partition, and different partitions can have identical rowIds.  I can't believe that I forgot that, but I wrote that code months ago.

I added this comment at the top of DuplicateRowFinder`:

```
/**
 * This class throws an exception if two adjacent rows have the same identity.
 * It compares both the rowId column as well as the writeRedistributionColumns,
 * because it is called before the page is separated by partition, and 
 * adifferent partitions can have identical the rowId columns.
 */
```